onnxruntime/core/providers/cuda/math/matmul_integer.cc

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "matmul_integer.h"
+#include "matmul_integer.cuh"
+#include "core/providers/cpu/math/matmul_helper.h"
+#include "core/providers/cuda/shared_inc/fpgeneric.h"
+#include "core/providers/cuda/cuda_allocator.h"
+#include "core/providers/common.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+ONNX_OPERATOR_TYPED_KERNEL_EX(
+    MatMulInteger,
+    kOnnxDomain,
+    10,
+    int8_t,
+    kCudaExecutionProvider,
+    KernelDefBuilder()
+        .TypeConstraint("T1", DataTypeImpl::GetTensorType<int8_t>())
+        .TypeConstraint("T2", DataTypeImpl::GetTensorType<int8_t>())
+        .TypeConstraint("T3", DataTypeImpl::GetTensorType<int32_t>()),
+    MatMulInteger<int8_t, int8_t>);
+
+template <>
+Status MatMulInteger<int8_t, int8_t>::PadMatrix(
+    int64_t row,
+    int64_t col,
+    int64_t align_size,
+    const int8_t*& src,
+    int64_t& pad_size,
+    IAllocatorUniquePtr<int8_t>& temp_mem_holder) const {
+  pad_size = align_size - col % align_size;
+  if (pad_size != align_size) {
+    temp_mem_holder = GetScratchBuffer<int8_t>(row * (col + pad_size));
+    ORT_RETURN_IF_ERROR(PadMatrixInLeadingDimension(src, temp_mem_holder.get(), row, col, pad_size));
+    src = temp_mem_holder.get();
+  } else {
+    pad_size = 0;
+  }
+
+  return Status::OK();
+}
+
+template <>
+Status MatMulInteger<int8_t, int8_t>::ComputeInternal(OpKernelContext* ctx) const {
+  auto a = ctx->Input<Tensor>(0);
+  auto b = ctx->Input<Tensor>(1);
+  ORT_ENFORCE(a != nullptr && b != nullptr);
+
+  MatMulComputeHelper helper;
+  ORT_RETURN_IF_ERROR(helper.Compute(a->Shape(), b->Shape()));
+  Tensor* Y = ctx->Output(0, helper.OutputShape());
+
+  const int8_t* a_ptr = a->template Data<int8_t>();
+  const int8_t* b_ptr = b->template Data<int8_t>();
+  int32_t* output_ptr = Y->template MutableData<int32_t>();
+
+  // validate zero points
+  const int8_t* a_offset = nullptr;
+  const int8_t* b_offset = nullptr;
+  if (has_a_zero_point_) {
+    auto a_zero_point = ctx->Input<Tensor>(2);
+    ORT_ENFORCE(IsScalarOr1ElementVector(a_zero_point),
+                "MatmulInteger : input1 zero point must be a scalar or 1D tensor of size 1");
```

Comment:
The spec says 
a_zero_point (optional) : T1
Zero point tensor for input 'A'. It's optional and default value is 0. It could be a scalar or a 1-D tensor, which means a per-tensor or per-row quantization. If it's a 1-D tensor, its number of elements should be equal to the number of rows of input 'A'.

This check is that there's only one element isn't it, so if the number of rows in input 'A' was > 1 shouldn't the number of elements here match that count?

Reply:
Yes, that's right. However, the QuantizeLinear [https://github.com/onnx/onnx/blob/master/docs/Operators.md#quantizelinear](url) doesn't support per-axis quantization in ONNX, which means there won't be a per-row/per-column MatMulInteger in practice. That's why our CPU also implements as this. 
We need to fix the QuantizeLinear and this implementation as well.